### ISSUE TITLE
Untyped variable and wildcard patterns are not guaranteed to match

### ISSUE THREAD

**Issue by stereotype441** [dart-lang/language#2619]
Consider the following code:
```dart
f(int i) {
  if (i case _) {
    print('matched');
  } else {
    print('unreachable');
  }
}
```
Can this code ever print `unreachable`?  It seems like it should be impossible.  But based on my reading of the spec, I think it could.  `_` is considered by type inference to be equivalent to `int _`, which means that the match will fail if the runtime type of `i` is not a subtype of `int`.  In a program that's not fully migrated to null safety, it's possible that `i` will have the value `null`.

This is really unfortunate, because even though this is unlikely to arise very often in practice, flow analysis has to reflect the semantics precisely (to avoid unsoundness escalalation), and therefore the `else` branch can't be flagged as unreachable code.

We could fix this by changing this line from the spec (from Runtime semantics > Matching > Variable):

> 2.  If the runtime type of `v` is not a subtype of `T` then the match fails.

To something like:

> 2. If the matched value type `M` is not a subtype of `T`, and the runtime type of `c` is not a subtype of `T`, then the match fails.

In other words, if the variable pattern looked like it was guaranteed to match based on static analysis, then it would be guaranteed to match.  And therefore all untyped variable and wildcard patterns would be guaranteed to match.

Note: there would be a negative consequence to that.  A program like this could print `i = null` in a program that's not fully migrated:

```dart
f(int i) {
  if (i case int x) {
    print('i = $x');
  }
}
```

I personally think that's ok.  IMHO it's no worse than the other examples of mixed mode unsoundness we have in Dart today, and it's way less bad than allowing `_` to sometimes fail to match.

**Comment by stereotype441** [comment]
CC @munificent @leafpetersen @jakemac53 @lrhn @mit-mit @natebosch @eernstg @kallentu
